File: go/objectmodel/schema.go
package objectmodel

import (
	"fmt"
	"slices"

	"gopkg.in/yaml.v3"
)

// primitiveSet is the eight irreducible atoms (SPEC §6.1). The order is the
// order SPEC §6.1 states them in, and it is also the canonical member order
// of a node (see canonical.go).
//
// D-003 was amended on 2026-05-04 to absorb D-011 and now says eight, not
// seven; the KB snapshot chunk c4255 still says seven. SPEC §6.1 is the
// authority here and it says eight.
var primitiveSet = []string{
	"shape", "role", "behavior", "contract", "address", "identity", "event", "access",
}

func isPrimitiveName(k string) bool { return slices.Contains(primitiveSet, k) }

// Shapes the vector schema language declares (conformance/README.md).
const (
	shapeScalar = "scalar"
	shapeList   = "list"
	shapeObject = "object"
	shapeOpaque = "opaque"
)

// schemaKeywords are the non-primitive keys of the vector schema language.
// `shape` is both a keyword and a primitive name: it is declared as a
// keyword (`shape: scalar` plus a sibling `scalar_type:`) and materialized
// as the `shape` primitive node (see primitives.go).
var schemaKeywords = map[string]bool{
	"shape": true, "scalar_type": true, "default": true, "required": true,
	"children": true, "item": true, "sealed_from": true, "content": true,
}

type sealedRef struct {
	template    string
	path        string
	hasTemplate bool
	hasPath     bool
}

type schemaNode struct {
	shape      string
	hasShape   bool
	scalarType string
	def        any
	hasDefault bool
	required   bool
	children   map[string]*schemaNode
	childOrder []string
	item       *schemaNode
	sealed     *sealedRef
	prims      map[string]any
	primOrder  []string
}

type templateEntry struct {
	node       *schemaNode
	content    any
	hasContent bool
}

// Schema is a loaded, schema-load-validated CIC schema.
type Schema struct {
	model     string
	root      *schemaNode
	templates map[string]map[string]*templateEntry
}

// LoadSchema parses and validates a schema. The checks it runs — INV-010,
// INV-015 and INV-005 — are the three the corpus places at stage
// `schema-load`, a stage SPEC §8 does not list (docs/spec-defects.md SD-002).
func LoadSchema(data []byte) (*Schema, error) {
	var raw any
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, newError(CodeMalformedDocument, "INV-013", StageSchemaLoad, "$",
			fmt.Sprintf("schema is not valid YAML: %v", err))
	}
	m, ok := asMap(raw)
	if !ok {
		return nil, newError(CodeMalformedDocument, "INV-013", StageSchemaLoad, "$",
			"schema must be a mapping")
	}

	s := &Schema{templates: map[string]map[string]*templateEntry{}}
	if mv, ok := m["model"]; ok && mv != nil {
		s.model = fmt.Sprint(mv)
	}

	rootRaw, ok := m["root"]
	if !ok {
		return nil, newError(CodeMalformedDocument, "INV-013", StageSchemaLoad, "$.root",
			"schema must declare a root node")
	}
	root, err := parseSchemaNode(rootRaw, "$")
	if err != nil {
		return nil, err
	}
	s.root = root

	if tRaw, ok := m["templates"]; ok && tRaw != nil {
		tm, ok := asMap(tRaw)
		if !ok {
			return nil, newError(CodeMalformedDocument, "INV-015", StageSchemaLoad, "$templates",
				"templates must be a mapping")
		}
		for _, name := range sortedKeys(tm) {
			pm, ok := asMap(tm[name])
			if !ok {
				return nil, newError(CodeMalformedDocument, "INV-015", StageSchemaLoad,
					"$templates."+name, "a template must be a mapping of path to node")
			}
			s.templates[name] = map[string]*templateEntry{}
			for _, p := range sortedKeys(pm) {
				entryPath := "$templates." + name + "." + p
				node, err := parseSchemaNode(pm[p], entryPath)
				if err != nil {
					return nil, err
				}
				entry := &templateEntry{node: node}
				if em, ok := asMap(pm[p]); ok {
					if c, ok := em["content"]; ok {
						entry.content, entry.hasContent = c, true
					}
				}
				s.templates[name][p] = entry
			}
		}
	}

	if err := s.schemaLoadChecks(); err != nil {
		return nil, err
	}
	return s, nil
}

func parseSchemaNode(raw any, path string) (*schemaNode, error) {
	m, ok := asMap(raw)
	if !ok {
		return nil, newError(CodeMalformedDocument, "INV-013", StageSchemaLoad, path,
			"a schema node must be a mapping")
	}
	n := &schemaNode{
		children: map[string]*schemaNode{},
		prims:    map[string]any{},
	}
	for _, k := range sortedKeys(m) {
		v := m[k]
		switch {
		case k == "shape":
			n.shape, n.hasShape = fmt.Sprint(v), true
		case k == "scalar_type":
			n.scalarType = fmt.Sprint(v)
		case k == "default":
			n.def, n.hasDefault = v, true
		case k == "required":
			b, _ := v.(bool)
			n.required = b
		case k == "content":
			// consumed by the template layer, not part of the node
		case k == "children":
			cm, ok := asMap(v)
			if !ok {
				return nil, newError(CodeMalformedDocument, "INV-013", StageSchemaLoad,
					path+".children", "children must be a mapping")
			}
			for _, cn := range sortedKeys(cm) {
				child, err := parseSchemaNode(cm[cn], path+"."+cn)
				if err != nil {
					return nil, err
				}
				n.children[cn] = child
				n.childOrder = append(n.childOrder, cn)
			}
		case k == "item":
			item, err := parseSchemaNode(v, path+"[]")
			if err != nil {
				return nil, err
			}
			n.item = item
		case k == "sealed_from":
			sm, ok := asMap(v)
			if !ok {
				return nil, newError(CodeSealedMissingTemplateOrPath, "INV-015", StageSchemaLoad,
					path+".sealed_from", "sealed_from must be a mapping carrying template and path")
			}
			ref := &sealedRef{}
			if t, ok := sm["template"]; ok && t != nil {
				ref.template, ref.hasTemplate = fmt.Sprint(t), true
			}
			if p, ok := sm["path"]; ok && p != nil {
				ref.path, ref.hasPath = fmt.Sprint(p), true
			}
			n.sealed = ref
		case isPrimitiveName(k):
			n.prims[k] = v
			n.primOrder = append(n.primOrder, k)
		case schemaKeywords[k]:
			// exhaustive above; kept for clarity
		default:
			// Fail closed. SPEC does not define the schema language, so an
			// unrecognised key cannot be assumed inert (docs/spec-defects.md SD-011).
			return nil, newError(CodeUnknownSchemaKey, "INV-021", StageSchemaLoad, path+"."+k,
				fmt.Sprintf("`%s` is neither a schema keyword nor a primitive name", k))
		}
	}
	return n, nil
}

func (s *Schema) schemaLoadChecks() error {
	if err := checkSchemaNode(s.root, "$"); err != nil {
		return err
	}
	for _, name := range sortedTemplateNames(s.templates) {
		paths := s.templates[name]
		for _, p := range sortedTemplatePaths(paths) {
			if err := checkSchemaNode(paths[p].node, "$templates."+name+"."+p); err != nil {
				return err
			}
		}
	}
	return nil
}

func checkSchemaNode(n *schemaNode, path string) error {
	// INV-015 — a sealed term must carry both template and path.
	if n.sealed != nil && (!n.sealed.hasTemplate || !n.sealed.hasPath) {
		missing := "template"
		if n.sealed.hasTemplate {
			missing = "path"
		}
		return newError(CodeSealedMissingTemplateOrPath, "INV-015", StageSchemaLoad,
			path+".sealed_from",
			fmt.Sprintf("sealed_from does not declare `%s`; template identity alone is not provenance", missing))
	}

	// INV-010 — a schema must not declare a child named `values` at a
	// structured object position. This is what makes INV-009 total.
	if n.shape == shapeObject {
		if _, ok := n.children["values"]; ok {
			return newError(CodeSchemaReservedChildValues, "INV-010", StageSchemaLoad,
				path+".values",
				"a schema must not declare a child property named `values` at a structured object position; it would collide with the envelope discriminator")
		}
	}

	// INV-005 — the primitive declaration graph must be acyclic.
	for _, name := range n.primOrder {
		next := []string{name}
		if err := checkPrimitiveCycle(n.prims[name], path+"."+name, next); err != nil {
			return err
		}
	}

	for _, c := range n.childOrder {
		if err := checkSchemaNode(n.children[c], path+"."+c); err != nil {
			return err
		}
	}
	if n.item != nil {
		if err := checkSchemaNode(n.item, path+"[]"); err != nil {
			return err
		}
	}
	return nil
}

// checkPrimitiveCycle enforces INV-005.
//
// INV-005 is worded as "a primitive's own primitives MUST NOT, directly or
// transitively, re-declare the node they hang from", but the vector schema
// language is a finite tree with no reference mechanism, so no schema written
// in it can re-declare a node and no materialization written against it can
// fail to terminate. invalid/008_cyclic_primitive_declaration therefore tests
// a different (stronger) property than INV-005 states: acyclicity of the
// primitive-*name* graph. This implementation enforces the property the
// vector requires and records the discrepancy as docs/spec-defects.md SD-005.
func checkPrimitiveCycle(v any, path string, stack []string) error {
	m, ok := asMap(v)
	if !ok {
		return nil
	}
	for _, k := range sortedKeys(m) {
		kp := path + "." + k
		if isPrimitiveName(k) {
			if slices.Contains(stack, k) {
				return newError(CodeCyclicPrimitiveDeclaration, "INV-005", StageSchemaLoad, kp,
					"the primitive declaration graph must be acyclic: a primitive must not directly or transitively re-declare the node it hangs from")
			}
			next := make([]string, len(stack)+1)
			copy(next, stack)
			next[len(stack)] = k
			if err := checkPrimitiveCycle(m[k], kp, next); err != nil {
				return err
			}
			continue
		}
		if err := checkPrimitiveCycle(m[k], kp, stack); err != nil {
			return err
		}
	}
	return nil
}

func sortedTemplateNames(t map[string]map[string]*templateEntry) []string {
	keys := make([]string, 0, len(t))
	for k := range t {
		keys = append(keys, k)
	}
	slices.Sort(keys)
	return keys
}

func sortedTemplatePaths(t map[string]*templateEntry) []string {
	keys := make([]string, 0, len(t))
	for k := range t {
		keys = append(keys, k)
	}
	slices.Sort(keys)
	return keys
}

File: rust/src/schema.rs
//! The schema language (`conformance/README.md`), and the `schema-load` stage.
//!
//! Three checks live here because the corpus places them here: INV-010 (a
//! schema may not declare a child called `values`), INV-015 (a `sealed_from`
//! carries both template and path) and the model-version check of INV-033.
//! `schema-load` is not a stage SPEC §8 lists — `docs/spec-defects.md` SD-002.

use crate::error::{code, Error, Result, Stage};
use crate::value::Value;
use crate::MODEL_VERSION;

/// The eight irreducible atoms, in the order SPEC §6.1 states them. That order
/// is also the canonical member order of a node, so this array is both the
/// membership test and the sort key.
pub const PRIMITIVES: [&str; 8] = [
    "shape", "role", "behavior", "contract", "address", "identity", "event", "access",
];

#[must_use]
pub fn is_primitive(name: &str) -> bool {
    PRIMITIVES.contains(&name)
}

/// Keys of the schema language that are not primitive names. `shape` is both:
/// declared as a keyword (`shape: scalar` with a sibling `scalar_type:`) and
/// materialized as the `shape` primitive node.
const KEYWORDS: [&str; 8] = [
    "shape",
    "scalar_type",
    "default",
    "required",
    "children",
    "item",
    "sealed_from",
    "content",
];

#[derive(Debug, Clone)]
pub struct SealedRef {
    pub template: String,
    pub path: String,
}

#[derive(Debug, Clone, Default)]
pub struct SchemaNode {
    pub shape: Option<String>,
    pub scalar_type: Option<String>,
    pub default: Option<Value>,
    pub required: bool,
    pub children: Vec<(String, SchemaNode)>,
    pub item: Option<Box<SchemaNode>>,
    pub sealed_from: Option<SealedRef>,
    /// Primitive declarations, in declaration order. Re-ordered to §6.1 order
    /// only at serialization.
    pub prims: Vec<(String, Value)>,
    /// `content` of a template entry: the values a sealed instance carries.
    pub content: Option<Value>,
}

impl SchemaNode {
    #[must_use]
    pub fn child(&self, name: &str) -> Option<&SchemaNode> {
        self.children
            .iter()
            .find(|(n, _)| n == name)
            .map(|(_, c)| c)
    }
}

#[derive(Debug, Clone)]
pub struct Schema {
    pub model: String,
    pub root: SchemaNode,
    /// template name -> path within the template -> node
    pub templates: Vec<(String, Vec<(String, SchemaNode)>)>,
}

impl Schema {
    #[must_use]
    pub fn template_entry(&self, template: &str, path: &str) -> Option<&SchemaNode> {
        self.templates
            .iter()
            .find(|(n, _)| n == template)
            .and_then(|(_, entries)| entries.iter().find(|(p, _)| p == path))
            .map(|(_, n)| n)
    }
}

/// Parse and validate a schema on its own.
///
/// # Errors
/// Returns the first violation found, at stage `schema-load`.
pub fn load(data: &[u8]) -> Result<Schema> {
    let raw = crate::value::parse(data, Stage::SchemaLoad, "$", "schema")?;
    let Some(m) = raw.as_map() else {
        return Err(Error::new(
            code::MALFORMED_DOCUMENT,
            "INV-013",
            Stage::SchemaLoad,
            "$",
            "schema must be a mapping",
        ));
    };

    // INV-033 — an object is handed over at a KNOWN model version, so a schema
    // at an unknown one cannot produce such an object and is refused here
    // rather than at the boundary, where the fault is not.
    let model = match m.get("model") {
        Some(Value::Str(s)) => s.clone(),
        Some(other) => other.to_plain_string(),
        None => {
            return Err(Error::new(
                code::UNSUPPORTED_MODEL_VERSION,
                "INV-033",
                Stage::SchemaLoad,
                "$.model",
                "schema declares no model version",
            ))
        }
    };
    if model != MODEL_VERSION {
        return Err(Error::new(
            code::UNSUPPORTED_MODEL_VERSION,
            "INV-033",
            Stage::SchemaLoad,
            "$.model",
            format!(
                "model version `{model}` is not supported; this implementation is {MODEL_VERSION}"
            ),
        ));
    }

    let Some(root_raw) = m.get("root") else {
        return Err(Error::new(
            code::MALFORMED_DOCUMENT,
            "INV-013",
            Stage::SchemaLoad,
            "$.root",
            "schema must declare a root node",
        ));
    };
    let root = parse_node(root_raw, "$")?;

    let mut templates = Vec::new();
    if let Some(t) = m.get("templates") {
        let Some(tm) = t.as_map() else {
            return Err(Error::new(
                code::MALFORMED_DOCUMENT,
                "INV-015",
                Stage::SchemaLoad,
                "$templates",
                "templates must be a mapping",
            ));
        };
        for (name, paths) in &tm.0 {
            let Some(pm) = paths.as_map() else {
                return Err(Error::new(
                    code::MALFORMED_DOCUMENT,
                    "INV-015",
                    Stage::SchemaLoad,
                    format!("$templates.{name}"),
                    "a template must be a mapping of path to node",
                ));
            };
            let mut entries = Vec::new();
            for (path, node) in &pm.0 {
                entries.push((
                    path.clone(),
                    parse_node(node, &format!("$templates.{name}.{path}"))?,
                ));
            }
            templates.push((name.clone(), entries));
        }
    }

    Ok(Schema {
        model,
        root,
        templates,
    })
}

fn parse_node(raw: &Value, path: &str) -> Result<SchemaNode> {
    let Some(m) = raw.as_map() else {
        return Err(Error::new(
            code::MALFORMED_DOCUMENT,
            "INV-013",
            Stage::SchemaLoad,
            path,
            "a schema node must be a mapping",
        ));
    };

    let mut n = SchemaNode::default();
    for (k, v) in &m.0 {
        match k.as_str() {
            "shape" => n.shape = Some(v.to_plain_string()),
            "scalar_type" => n.scalar_type = Some(v.to_plain_string()),
            "default" => n.default = Some(v.clone()),
            "required" => n.required = matches!(v, Value::Bool(true)),
            "content" => n.content = Some(v.clone()),
            "children" => n.children = parse_children(v, path)?,
            "item" => n.item = Some(Box::new(parse_node(v, &format!("{path}[]"))?)),
            "sealed_from" => {
                let sealed_path = format!("{path}.values.sealed_from");
                let Some(sm) = v.as_map() else {
                    return Err(Error::new(
                        code::SEALED_MISSING_TEMPLATE_OR_PATH,
                        "INV-015",
                        Stage::SchemaLoad,
                        sealed_path,
                        "sealed_from must be a mapping carrying template and path",
                    ));
                };
                // INV-015 — both, always. Template identity alone is not
                // provenance: it says which template, not which node in it.
                //
                // `null` counts as absent. A key present with a null value is
                // Some(Value::Null) and would otherwise pass a presence test,
                // producing a sealed origin whose template is the string
                // "null" — provenance pointing at nothing, which is worse than
                // a rejection because it looks like an answer.
                let present = |k: &str| match sm.get(k) {
                    None | Some(Value::Null) => None,
                    Some(v) => Some(v),
                };
                let (Some(t), Some(p)) = (present("template"), present("path")) else {
                    return Err(Error::new(
                        code::SEALED_MISSING_TEMPLATE_OR_PATH,
                        "INV-015",
                        Stage::SchemaLoad,
                        sealed_path,
                        "a sealed_from must carry both template and path",
                    ));
                };
                n.sealed_from = Some(SealedRef {
                    template: t.to_plain_string(),
                    path: p.to_plain_string(),
                });
            }
            other if is_primitive(other) => n.prims.push((other.to_string(), v.clone())),
            other => {
                return Err(Error::new(
                    code::UNKNOWN_SCHEMA_KEY,
                    "INV-021",
                    Stage::SchemaLoad,
                    path,
                    format!("`{other}` is neither a schema keyword nor a primitive name"),
                ))
            }
        }
    }

    // A node declares its shape, or is sealed from a template that declares it.
    if n.shape.is_none() && n.sealed_from.is_none() {
        return Err(Error::new(
            code::SCHEMA_SHAPE_MISSING,
            "INV-011",
            Stage::SchemaLoad,
            path,
            "a schema node must declare a shape",
        ));
    }
    let _ = KEYWORDS;
    Ok(n)
}

/// The `children` mapping of a schema node.
///
/// Split out of `parse_node` because two of the reservations below carry their
/// own explanation and the arm had grown past the point where the reader can
/// see the whole match.
fn parse_children(v: &Value, path: &str) -> Result<Vec<(String, SchemaNode)>> {
    let Some(cm) = v.as_map() else {
        return Err(Error::new(
            code::MALFORMED_DOCUMENT,
            "INV-013",
            Stage::SchemaLoad,
            format!("{path}.children"),
            "children must be a mapping",
        ));
    };
    let mut out = Vec::new();
    for (cname, cnode) in &cm.0 {
        // INV-010 — `values` and, since 0.2, `origin` are reserved members of
        // every node. A child of either name would be indistinguishable from
        // the member itself.
        if cname == "values" || cname == "origin" {
            return Err(Error::new(
                code::SCHEMA_RESERVED_CHILD_VALUES,
                "INV-010",
                Stage::SchemaLoad,
                format!("{path}.values.{cname}"),
                format!("a schema may not declare a child named `{cname}`; it is a reserved member of every node"),
            ));
        }
        out.push((
            cname.clone(),
            parse_node(cnode, &format!("{path}.values.{cname}"))?,
        ));
    }
    Ok(out)
}
